bpo-29729: uuid.UUID now accepts bytes-like object

vstinner opened the pull request:
uuid.UUID doesn't require the 'bytes' argument to be an instance of
bytes: accept bytes-like types, bytearray and memoryview for example.

<!-- issue-number: bpo-29729 -->
https://bugs.python.org/issue29729
<!-- /issue-number -->

vstinner committed:
bpo-29729: uuid.UUID now accepts bytes-like object

uuid.UUID doesn't require the 'bytes' argument to be an instance of
bytes: accept bytes-like types, bytearray and memoryview for example.

serhiy-storchaka reviewed (commented):
Add a "versionchanged" directive in the module documentation.

Hmm, `memoryview` is not accepted as `bytes_le`.  And the length check works correctly only with memoryview with byte format.

vstinner committed:
Accept also bytes-like for bytes_le

* Accept also bytes-like for bytes_le: convert memoryview to bytes to
  use slices. Check also the length after the conversion to integer
  to raise a TypeError even if the length is not 16.
* Fix unit tests: pass bytes to 'bytes' and 'bytes_le' parameters
* Add more tests on Unicode passed to bytes and bytes_le
* Document UUID changes

vstinner commented:
> Hmm, memoryview is not accepted as bytes_le

Oh, I didn't look at bytes_le. I also modified bytes_le to accept bytes-like objects.

> And the length check works correctly only with memoryview with byte format.

I took care of such issues in my updated PR.

serhiy-storchaka commented on Lib/uuid.py:
```diff
@@ -159,15 +159,18 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
                 raise ValueError('badly formed hexadecimal UUID string')
             int = int_(hex, 16)
         if bytes_le is not None:
+            if not isinstance(bytes_le, (bytes_, bytearray)):
```
Add `int`.

vstinner replied:
I don't understand why bytes_le should accept an int? Do you have an example please?

serhiy-storchaka replied:
It should not! But accepts. Add a test for `bytes_le=16`. :wink:

serhiy-storchaka commented on Misc/NEWS.d/next/Library/2017-09-28-16-21-13.bpo-29729.skDkJA.rst:
```diff
@@ -0,0 +1,3 @@
+uuid.UUID now accepts bytes-like object. The constructor doesn't require the
+'bytes' argument to be an instance of bytes: accept bytes-like types,
```
... and 'bytes_le'

vstinner replied:
Oops, fixed.

serhiy-storchaka commented:
A ValueError is raised if 'bytes_le' is a str of len != 16. To solve this duplicate `int.from_bytes()` for `bytes_le` instead of reusing the code for `bytes`.

vstinner commented:
> A ValueError is raised if 'bytes_le' is a str of len != 16.

I added an unit tests to ensure that a *TypeError* is raised if you pass a str of any length to bytes or bytes_le.

vstinner committed:
Update the NEWS entry.

vstinner commented:
Please don't merge my PR until PR #3796 is merged. IMHO PR #3796 is more important :-)

serhiy-storchaka commented:
> I added an unit tests to ensure that a *TypeError* is raised if you pass a str of any length to bytes or bytes_le.

Ah, yes, a bytes constructor raises an error.

But still there is a problem with memoryview with non-byte format. `bytes_le=memoryview(b'\0'*16).cast('I')` works, but `bytes=memoryview(b'\0'*16).cast('I')` is rejected and `bytes=memoryview(b'\0'*64).cast('I')` is accepted by mistake.

vstinner committed:
Merge branch 'master' into uuid_bytes

vstinner committed:
Fix UUID(bytes_le=16)

vstinner commented:
PR #3796 has been merged. I merged master into my branch.

> But still there is a problem with memoryview with non-byte format. memoryview(b'\0'*64).cast('I') is accepted by mistake.

I don't think that we should overengineer the code here. We are consenting adults here. You can also find a value to skip input validations. It's really hard to write the perfect validation. If you can a random memoryview, yeah, maybe we get a valid UUID object whereas you shouldn't. Well, I don't think that we should detect this corner case. Why would you do that on purpose?

vadmium commented on Lib/uuid.py:
```diff
@@ -160,15 +160,19 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
+            # Don't cast int to bytes to get a TypeError above
+            if not isinstance(bytes_le, (bytes_, bytearray, int_)):
+                bytes_le = bytes_(bytes_le)
             if len(bytes_le) != 16:
```
By _above_ in the comment, do you mean this check, which is actually below the comment? :)

vstinner replied:
I removed the comment.

vadmium reviewed (commented):
It may not be worth supporting every kind of invalid input (unless you have to handle untrusted input). But it is important to implement what the documentation promises (or fix the documentation). As of rev a169e89, _bytes_ has to also support _len_.

vstinner committed:
Remove a comment

vstinner commented:
@vadmium: "It may not be worth supporting every kind of invalid input (unless you have to handle untrusted input). But it is important to implement what the documentation promises (or fix the documentation). As of rev a169e89, bytes has to also support len."

I'm sorry, but I don't understand what you are expecting from me. What do you mean by "bytes has to also support len"? Do you expect a nicer error message if len() fails?

vstinner commented:
@serhiy-storchaka, @vadmium: Would you mind to review the latest version of my change, please?

vadmium commented:
For your code to work, the object passed as the _bytes_ argument has to implement `__len__` returning 16. But your documentation promises any bytes-like object should work (as long as it is a string of 16 bytes). Since the glossary definition of bytes-like objects has no requirements about implementing `__len__`, I would expect _len_ to not be used on the argument. You could change the code for _bytes_ to make a copy like you do for _bytes_le_:

``` python
bytes = bytes_(bytes)  # Convert bytes-like to bytes object
if len(bytes) != 16 . . .
```

Or you could use _memoryview_:

``` python
with memoryview(bytes) as view:
    if view.nbytes != 16 . . .
```

Or you could change the documentation:

*bytes_le* now accepts any bytes-like object, and *bytes* now accepts any bytes-like object for which _len_ returns 16.

vstinner committed:
Merge branch 'master' into uuid_bytes

vstinner committed:
Repeat that objects must have a length of 16

vstinner commented:
> Or you could change the documentation:

UUID constructor documentation is explicit:

```
   Create a UUID from either a string of 32 hexadecimal digits, a string of 16
   bytes as the *bytes* argument, a string of 16 bytes in little-endian order as
   the *bytes_le* argument, ...
```

I guess that your remarks was on my "versionchanged" markup and the NEWS entry. I added "of 16 bytes" there.

pitrou commented on Lib/uuid.py:
```diff
@@ -160,15 +160,18 @@ def __init__(self, hex=None, bytes=None, bytes_le=None, fields=None,
+            int = int_.from_bytes(bytes, byteorder='big')
+            # test length after the conversion to raise a TypeError exception
+            # if 'bytes' type is str even if 'bytes' length is not 16
             if len(bytes) != 16:
```
```
>>> a = array.array('i', [-1]*4)
>>> x = int.from_bytes(a, byteorder='big')
>>> len(a)
4
>>> x.bit_length()/8
16.0
```

And conversely:
```
>>> a = array.array('i', [-1]*16)
>>> x = int.from_bytes(a, byteorder='big')
>>> len(a)
16
>>> x.bit_length()/8
64.0
```

vadmium commented:
Yes I was talking about the _versionchanged_ note, which now says "any bytes-like objects of 16 bytes". But I don't think this is clear enough. Antoine and Serhiy already gave examples of bytes-like objects containing exactly 16 bytes where _len_ returns a different number.

Since you already copy _bytes_le_ into a true bytes object, why not parallel this for the _bytes_ parameter?

``` python
if bytes is not None:
    # Check for int so that bytes=16 triggers an exception below
    if not isinstance(bytes, (bytes_, bytearray, int_)):
        bytes = bytes_(bytes)  # Copy any bytes-like object
    if len(bytes) != 16:
        raise ValueError('bytes is not a 16-byte string')
    int = int_.from_bytes(bytes, 'big')
```

vstinner commented:
I decided to work on https://bugs.python.org/issue29729 because I understand that the proposed change was just to remove an assertion. That's all.

It seems like you expect much more changes, much more complex code. It's already my 6th version of my pull request.

Honestly, I don't care much of the uuid module. I don't think that anyone ever tried to call UUID(bytes=value) with a type different than bytes (ok, apart the bug reporter ;-)). If I would really need to pass non-bytes, having to cast manually with bytes(value) wouldn't burn my hands.

For all these reasons, I abandon my change. Feel free to reuse my code if someone wants to pursue the effort on UUID enhancement ;-)

serhiy-storchaka commented:
The problem with adding support of bytes-like objects is that in different cases this term has different meaning. The definition in the glossary is too wide, in many particular cases we need the support of not just the buffer protocol, but len(), indexing, slicing, iterating, startswith() or even lower().

I think we need different terms for different levels of "bytesness".

vstinner commented:
I think that we should replace the assertion with a regular if and raise a proper TypeError, rather than trying to support "bytes-like" objects. It's fine to require bytes on UUID. UUID is not a common type, and passing bytes to UUID is even less common (IMHO).
